Botan: First Milestone blog (#4)
haskell-cryptography/haskell-cryptography.github.io: content/blog/botan-first-milestone.md

Reviewer: You have more than two `<h1>` headers, this breaks the semantics of HTML

template = "blog/page.html"

[extra]
lead = "After 8 months of work, we've reached an important milestone. Now we're here to tell you about what is coming next."
+++

# Botan: The First Milestone

First off, huge thanks to [Mercury][mercury] for funding the [first proposal][first proposal] and helping us reach this milestone - this library would not be in this state without their support. If you are an engineer looking for a savvy place to work, I hear they are [hiring][mercury hiring]!

With [botan-bindings][botan-bindings] and [botan-low][botan-low] having reached their initial `0.0.1` release and `botan` making it to [package candidate][botan] status, we've reached an important milestone - the first released version!

After 8 months of work, celebrations were in order, and after a bit of a breather, we're back in action to tell you about what is coming next.

But first...

# A call for users

Do you use one of the following libraries?

- `crypton` / `cryptonite`
- `libsodium`
- `saltine`

What are you using cryptography for? Would you be interested in trying something new? Perhaps an alternate backend for your cryptography needs?

> We're working on `botan`, a cryptographic kitchen sink, and could use your feedback! We've successfully published `botan-bindings` and `botan-low`, and now are hard at work on getting `botan` ready too. That's where you come in! We can't listen to your feedback if there isn't any.

Reviewer: This sentence could actually be the opening sentence. The blog post opens with a shout-out to Mercury and then talks about milestones, but you don't explain what it is until after.

I am Leo Dillinger, a member of [Haskell Cryptography Group][haskell cryptography group], and I am working with the [Haskell Foundation][haskell foundation] to develop free and open-source software for you and the Haskell ecosystem.

# What are our goals?

We seek to provide trusted, open-source cryptography solutions to you. Much of the existing Haskell cryptography ecosystem is aging, unmaintained and unaudited, or very limited in scope, we are seeking to improve that.

# What is Botan?

Botan is an comprehensive, open-source, BSD-licenced, C++ cryptography library with a stable C API. It offers a broad variety of functionality and algorithms, including **post-quantum cryptography**, is developed and maintained by an active community, and has been [audited][botan audit] in the past.

By binding to Botan, we have solved a significant problem of providing much of the necessary 'cryptographic kitchen sink' via a suitably performant, suitably licensed, open-source library. Furthermore, we do this without imposing a large maintenance burden on the Haskell community, as we are not required to maintain the Botan cryptography library itself, only the bindings to it.

See the

> [first proposal][first proposal] for more details.

Reviewer: Yeah this should definitely come first.

We'd also like to work on the `MonadRandomIO / NonDeterministic` monad in order to integrate with `random` and take advantage of the `Uniform` class to make sampling / generation of random data easy.

We'd like to also take advantage of the secure memory erasure by providing functions for creating temporary objects that are automatically zeroed when they go out of scope. This is a big one, we don't want sensitive data hanging around in memory.

Finally, we'd also like to continue improving, unifying, and standardizing the `botan-low` bindings for consistency overall, as the techniques for binding to functions were refined over time, and have not been evenly applied.

---

These are some of the things that we'd like to accomplish over the coming months.

You can find more details in the updated [second funding proposal][second proposal]. This proposal is a continuation of the efforts of the first proposal, and is motivated by the same long-term goals.

# A flag planted on the horizon

As part of the Haskell Cryptography Group, we're about more than just maintenance of existing libraries - we want to develop a full suite of modern cryptography libraries.

> Cryptography is a different niche, but we see the success of Haskell web server libraries as an example of a healthy, community-driven ecosystem. There are packages ranging from the low-level like `wai` supporting multiple backends from the swiftly simple `scotty` to the deeply complex `servant`, frontends like `blaze-html` and `lucid` - a rich set of libraries that provide a flexible enough set of solutions at whatever level of abstraction you need, that take full advantage of Haskell's powerful type system.

Reviewer: Not too sure what's the relationship with HTML generation I have to admit

+++
title = "Botan: The first milestone"
description = "A step towards improving the state of Haskell cryptography"
date = 2024-06-03T12:15:00
updated = 2024-06-03T12:15:00
draft = false # Leave this as true so that it is not published
template = "blog/page.html"

[extra]
lead = "After 8 months of work, we've reached an important milestone. Now we're here to tell you about what is coming next."
+++

# Botan: The First Milestone

What are you using cryptography for? Would you be interested in trying something new? Perhaps an alternate backend for your cryptography needs?

We're working on `botan`, a cryptographic kitchen sink, and after 8 months of work, we've reached an important milestone: we've successfully published the first released version `0.0.1` of  [botan-bindings][botan-bindings] and [botan-low][botan-low], and `botan` has made it to [package candidate][botan] status!

Now we are hard at work on getting the first release version of `botan` ready too, and could use your feedback!

> ## What is Botan?

Reviewer: I would suggest that this sub-header be removed and that its content appears right next to `# Botan: the first milestone".

Author: I've done my best to juggle things around again, but it conflicts with your suggestion of placing the "What are you using cryptography for?" hook at the beginning. I've followed the following logic for the revised first section:

- Hook has to come first, its the definition of a hook
- Author intro  has to happen immediately after the hook, it would be first otherwise
- Summary ("What is this blog about? It is about botan's milestone) introduces topic (milestone) and subject (botan)
- Explanation ("What is botan") has to come after summary and topic / subject introductions
- "Summarize, then Explain" is an important rule for me because otherwise I easily become impossible to follow

Effectively it embodies the following interaction:

"Are you looking for ____? I am ____ and I would like to talk to you about how ____ can help. What is _____? _____ is ...."

Botan is an comprehensive, open-source, BSD-licenced, C++ cryptography library with a stable C API. It offers a broad variety of functionality and algorithms, including **post-quantum cryptography**, is developed and maintained by an active community, and has been [audited][botan audit] in the past.

By binding to Botan, we have solved a significant problem of providing much of the necessary 'cryptographic kitchen sink' via a suitably performant, suitably licensed, open-source library. Furthermore, we do this without imposing a large maintenance burden on the Haskell community, as we are not required to maintain the Botan cryptography library itself, only the bindings to it.

## A call for users

Do you use one of the following libraries?

- `crypton` / `cryptonite`
- `libsodium`

> - `saltine`

Author: Though it is a long term goal to provide a similarly high-level (but not necessarily libsodium-compatible) interface focusing on simplified operations with that use recommended algorithms, I see what you mean, and we aren't quite there yet. Best speak to our direct audience now, and we'll get back to that later at a more appropriate time.

That's where you come in! We'd like you to consider giving [botan][botan github] a try!

> `botan-low` is surprisingly viable out-of-box, and `botan` will be having it's own `0.0.1` release soon enough. We are interested in seeing how they perform in the real world, and we can only do that with users - we can't listen to your feedback if there isn't any!

Reviewer: ```suggestion
`botan-low` is viable out-of-box, and `botan` will be having it's own `0.0.1` release soon enough. We are interested in seeing how they perform in the real world, and we can only do that with users - we can't listen to your feedback if there isn't any!
```

Don't be surprised by your own achievements

- `amazonka-core`
- `hpack`
- `scotty`
- `servant`
- `stack`
- `tls`
- `x509`
- `warp`
- `websockets`

There's some pretty popular libraries on that list there, and their list of transitive dependents is quite large. These libraries sit at the root of a lot of production haskell code, and anything involving networking, APIs, and the internet is likely to depend on at least one of them. 

> Would it surprise you that they are all directly or indirectly dependent on `crypton`, which contains unaudited C implementations that must be maintained by the Haskell community as a fork after `cryptonite` was abandoned by its original author.

Author: Note: This may be a bit blunt. I don't necessarily want to say 'abandoned' but it is difficult not to after what happened. Is there a better way we can say this while still strongly projecting that essentially no new development is going to occur, just maintenance? (Eg, no new algorithms, which is particularly problematic in light of post-quantum cryptography)

> There's some pretty popular libraries on that list there, and their list of transitive dependents is quite large. These libraries sit at the root of a lot of production haskell code, and anything involving networking, APIs, and the internet is likely to depend on at least one of them. 

Reviewer: ```suggestion
These are popular libraries, and their list of transitive dependents is quite large. They sit at the root of a lot of production haskell code, and anything involving networking, APIs, and the internet is likely to depend on at least one of them. 
```

Would it surprise you that they are all directly or indirectly dependent on `crypton`, which contains unaudited C implementations that must be maintained by the Haskell community as a fork after `cryptonite` was abandoned by its original author.

This is a problem, and we're trying to change that with bindings to `botan`.

Having a solid, reliable, and well-maintained cryptography libraries is a huge benefit to the Haskell ecosystem in general, and is essential for most commercial and industrial use cases.

What are you using cryptography for? Let us know in the comments or with upvotes.

## Who are we?

I am Leo Dillinger, a member of [Haskell Cryptography Group][haskell cryptography group], and I am working with the [Haskell Foundation][haskell foundation] to develop free and open-source software for you and the Haskell ecosystem.

## What are our goals?

We seek to provide trusted, open-source cryptography solutions to you. Much of the existing Haskell cryptography ecosystem is aging, unmaintained and unaudited, or very limited in scope, we are seeking to improve that.

> Today, we're here to tell you about what is coming next.

Reviewer: It's a bit redundant, the cryptography group already introduced itself in a previous blog post. 

You should however introduce yourself much earlier in the blog post, possibly in the first sentence.

## The second milestone

That is what this next milestone is about - listening to feedback, improving the user experience, and seeing where the pain points are. Here's what we've heard, and here's what we're planning for the next three months.

### Improved installation support

One of the biggest pieces of feedback that we've received is the need for improved support for the installation of the `botan3` C++ library. This was a recurring item, and we've heard you loud and clear.

We'd like to spend a good chunk of time improving the installation process, with a specific focus on Windows[^1] and Linux support, working on the CI + unit tests for improved reliability and reporting overall.

[^1]: Definitely more highly requested than anticipated

> We're also looking into using `build-type: Configure` for bundling Botan C++ as a Haskell package for easy installation on all operating systems - we'd like for usage to be as easy as adding `botan` to your dependencies.

Reviewer: Wrong idea, especially with [HSEC-2024-002](https://github.com/haskell/security-advisories/pull/157/files#diff-269d148a2ac37251423165ca20885037feef75be318ae44cafd06d7d51f343dcR56-R58) that just came out.

Author: Bundled is the wrong term then - pardon my colloquialism; I don't mean actually including source code (I see the obvious flaw in that, and you've warned against it elsewhere) but some way fetching and building the latest version of botan via cabal if it is not already installed. I'll clarify this.

Reviewer: Ah I think I see what you mean. Perhaps something like this: https://dashbit.co/blog/rustler-precompiled ?

Reviewer 2: > but some way fetching and building the latest version of botan via cabal if it is not already installed.

This breaks:

- reproducibility (potentially)
- caching and offline installations (you shouldn't need a network connection to install a downloaded cabal package)
